driver-core/src/test/unit/com/mongodb/internal/connection/InternalStreamConnectionInitializerSpecification.groovy

```diff
@@ -243,6 +256,94 @@ class InternalStreamConnectionInitializerSpecification extends Specification {
+        def description = initializer.initialize(internalConnectionForSpeculativeAuthentication)
+
+        then:
+        description
+        1 * authenticator.authenticate(internalConnectionForSpeculativeAuthentication, _)
+        1 * authenticator.createSpeculativeAuthenticateCommand(internalConnectionForSpeculativeAuthentication)
+    }
+
+    @IgnoreIf({ !serverVersionAtLeast(4, 3) })
+    def 'should speculatively authenticate with SCRAM-SHA-1 authenticator'() {
+        given:
+        def credential = new MongoCredentialWithCache(createScramSha1Credential('user', 'database', 'pencil' as char[]))
+        def authenticator = Spy(ScramShaAuthenticator, constructorArgs: [credential, { 'fyko+d2lbbFgONRv9qkxdawL' }, { 'pencil' }])
+        def initializer = new InternalStreamConnectionInitializer(authenticator, null, [])
+
+        when:
+        enqueueSpeculativeAuthenticationResponsesForScramSha1()
+        def description = initializer.initialize(internalConnectionForSpeculativeAuthentication)
+
+        then:
+        description
+        1 * authenticator.authenticate(internalConnectionForSpeculativeAuthentication, _)
+        1 * authenticator.createSpeculativeAuthenticateCommand(internalConnectionForSpeculativeAuthentication)
+    }
+
+    @IgnoreIf({ !serverVersionAtLeast(4, 3) })
+    def 'should speculatively authenticate with X509 authenticator'() {
+        given:
+        def credential = new MongoCredentialWithCache(createMongoX509Credential())
+        def authenticator = Spy(X509Authenticator, constructorArgs: [credential])
+        def initializer = new InternalStreamConnectionInitializer(authenticator, null, [])
+
+        when:
+        enqueueSpeculativeAuthenticationResponsesForX509()
+        def description = initializer.initialize(internalConnectionForSpeculativeAuthentication)
+
+        then:
+        description
+        1 * authenticator.authenticate(internalConnectionForSpeculativeAuthentication, _)
+        1 * authenticator.createSpeculativeAuthenticateCommand(internalConnectionForSpeculativeAuthentication)
+    }
+
+    @IgnoreIf({ !serverVersionAtLeast(4, 3) })
+    def 'should not speculatively authenticate with Plain authenticator'() {
+        given:
+        def credential = new MongoCredentialWithCache(createPlainCredential('user', 'database', 'pencil' as char[]))
+        def authenticator = Spy(PlainAuthenticator, constructorArgs: [credential])
+        def initializer = new InternalStreamConnectionInitializer(authenticator, null, [])
+
+        when:
+        enqueueSpeculativeAuthenticationResponsesForPlain()
+        def description = initializer.initialize(internalConnectionForSpeculativeAuthentication)
+
+        then:
+        description
+        1 * authenticator.authenticate(internalConnectionForSpeculativeAuthentication, _)
+        authenticator.createSpeculativeAuthenticateCommand(internalConnectionForSpeculativeAuthentication) == null
```

Comment:
This assertion doesn't make sense: a PlainAuthenticator is not a SpeculativeAuthenticator doesn't have this method.  The only reason this works is because Spock is forgiving and just returns null for undefined methods on spies.

Reply:
I still would like to see this changed, as the assertion is not succeeding for the reason you want an assertion to succeed.  To test that the initializer works with a non-speculative authenticator, it just needs to assert that ismaster doesn't contains a speculativeAuthenticate field.

Reply:
Ok, I see my mistake.  I didn't realize until now that `SaslAuthenticator` implements `SpeculativeAuthenticator`, instead of just the subclasses that are actually able to speculatively authenticate.  Given that design, this test makes sense.